iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -221,30 +221,207 @@ struct MMTKernel {
+  kernel.k0 = 8;
+  kernel.n0 = 1;
+  kernel.lhsRegSize = 16;  // LHS NEON register type: int8x16
+  kernel.rhsRegSize = 8;   // RHS NEON register type: int8x8
+  kernel.accRegSize = 4;   // Accum NEON register type: int32x4
+  kernel.lhsRegs = 4;      // = 8x8/16 for 8x8 LHS elems, 16 per register
+  kernel.rhsRegs = 1;
+  kernel.accRegs = 2;  // = 8/4 for 8 accumulators, 4 per register
+  kernel.asmImpl = R"ASM(
+    // This kernel multiplies int8 values into temporary int16 values in
+    // registers v8--v15, then performs additions. We can't use
+    // multiply-accumulate instructions here because of the lack of an
+    // instruction multiplying int8 values and accumulating into int32, and
+    // we prefer to avoid the overhead of sign-extending the inputs from int8
+    // to int16 in this matrix*vector kernel where the largest matrix is the
+    // LHS.
+    ins v15.d[1], $(rhs:0).d[0]  // copy 1st half of $(rhs:0) to 2nd half of v15
+    smull v8.8h, $(lhs:0).8b, $(rhs:0).8b
+    smull2 v9.8h, $(lhs:0).16b, v15.16b
+    smull v10.8h, $(lhs:1).8b, $(rhs:0).8b
+    smull2 v11.8h, $(lhs:1).16b, v15.16b
+    smull v12.8h, $(lhs:2).8b, $(rhs:0).8b
+    smull2 v13.8h, $(lhs:2).16b, v15.16b
+    smull v14.8h, $(lhs:3).8b, $(rhs:0).8b
+    smull2 v15.8h, $(lhs:3).16b, v15.16b
+    // Now if we were able to codegen not just this MMT in isolation but
+    // a whole loop, we would diverge at this point: instead of doing the full
+    // additive reduction that the instructions below do, we would do only
+    // minimal reductions to temporary int32 accumulators
+    // (e.g. sadalp tmp.4s, v8.8h) and we would defer the rest of the work
+    // to the end of the loop. This is an example of how "MMT vector.contract"
+    // is not a perfect abstraction for "basic block of a MMT inner loop".
+    // Anyway...
+    //
+    // pairwise additions of int16 lanes to int32.
+    // So each result int32 is the sum of 2 products.
+    saddlp v8.4s, v8.8h
+    saddlp v9.4s, v9.8h
+    saddlp v10.4s, v10.8h
+    saddlp v11.4s, v11.8h
+    saddlp v12.4s, v12.8h
+    saddlp v13.4s, v13.8h
+    saddlp v14.4s, v14.8h
+    saddlp v15.4s, v15.8h
+    // pairwise additions of int32s, so each result is the sum of 4 products.
+    addp v8.4s, v8.4s, v9.4s
+    addp v10.4s, v10.4s, v11.4s
+    addp v12.4s, v12.4s, v13.4s
+    addp v14.4s, v14.4s, v15.4s
+    // pairwise additions of int32s, so each result is the sum of 8 products.
+    addp v8.4s, v8.4s, v10.4s
+    addp v12.4s, v12.4s, v14.4s
+    // Add to destination accumulators
+    add $(acc:0).4s, $(acc:0).4s, v8.4s
+    add $(acc:1).4s, $(acc:1).4s, v12.4s
+    )ASM";
+  kernel.asmClobbers = "~{v8},~{v9},~{v10},~{v11},~{v12},~{v13},~{v14},~{v15}";
+  return kernel;
+}
 
 // i8*i8->i32 kernel for Aarch64 NEON +dotprod
+//
+// This kernel is needed because: at the moment, codegen doesn't know how to
+// make use of dotprod instructions.
 MMTKernel MMTKernel_8x4x8_i8i8i32_Aarch64Dotprod_InlineAsm() {
```

Comment:
Adding a TODO here, which explains why it's not immediately needed (so I would like to do it in a separate PR, but you're right that it's important to do soon):
```
    // TODO: add a "kernel benefit" system whereby if two kernels are available
    // for the same shape and same data types, the fastest one (ie the one
    // using the most powerful available SIMD instructions) is selected.
    // This is incidentally not needed at the moment because currently no two
    // kernels share the exact same shape and data types.
```